crates/ide-diagnostics/src/handlers/trait_impl_reduntant_assoc_item.rs

```diff
@@ -0,0 +1,88 @@
+use hir::{db::ExpandDatabase, Const, Function, HasSource, TypeAlias};
+use syntax::ast;
+
+use crate::{Diagnostic, DiagnosticCode, DiagnosticsContext};
+
+// Diagnostic: trait-impl-reduntant-assoc_item
+//
+// Diagnoses reduntant trait items in a trait impl.
+pub(crate) fn trait_impl_reduntant_assoc_item(
+    ctx: &DiagnosticsContext<'_>,
+    d: &hir::TraitImplReduntantAssocItems,
+) -> Diagnostic {
+    let name = d.assoc_item.0.clone();
+    let assoc_item = d.assoc_item.1;
+    let db = ctx.sema.db;
+
+    let root = ctx.sema.db.parse_or_expand(d.file_id);
+    let range = root.text_range();
+    let impl_ = d.impl_.to_node(&root);
+    let trait_name = if let Some(trait_name) = trait_name(ctx, impl_) {
+        trait_name
+    } else {
+        String::from("{unkown}")
+    };
+
+    let (reduntant_item_name, diagnostic_range) = match assoc_item {
+        hir::AssocItem::Function(id) => (
+            format!("`fn {}`", name.display(db)),
+            Function::from(id).source(db).map(|it| it.syntax().value.text_range()).unwrap_or(range),
+        ),
+        hir::AssocItem::Const(id) => (
+            format!("`const {}`", name.display(db)),
+            Const::from(id).source(db).map(|it| it.syntax().value.text_range()).unwrap_or(range),
+        ),
+        hir::AssocItem::TypeAlias(id) => (
+            format!("`type {}`", name.display(db)),
+            TypeAlias::from(id)
+                .source(db)
+                .map(|it| it.syntax().value.text_range())
+                .unwrap_or(range),
+        ),
+    };
+
+    Diagnostic::new(
+        DiagnosticCode::RustcHardError("E0407"),
+        format!("{reduntant_item_name} is not a member of trait `{trait_name}`"),
+        diagnostic_range,
+    )
```

Comment:
Take a look at the `trait_impl_missing_assoc_item` assist here, you should be able to structure this similarly to that which is a bit more ocncise (that is the `diagnostic_range` fetching mainly)

Reply:
for `trait_impl_missing_assoc_item` diagnostic, the `diagnostic_range` just cover the trait name like the following:

https://github.com/rust-lang/rust-analyzer/blob/c9d189d1375e59a6c9b4d62fdede94ade001f6ee/crates/ide-diagnostics/src/handlers/trait_impl_missing_assoc_item.rs#L67-L71

while for `trait_impl_redundant_assoc_item` diagnostic, we want to cover every redundant assoc item inside the impl body, so the `diagnostic_range` fetching way of `trait_impl_missing_assoc_item` maynot fit here